Sync main with develop

pelesh opened the pull request:
Sync `main` branch with `develop` ahead of the release tag.

pelesh committed:
Merge branch 'cmake_cleanup' into 'develop'

Clean up of CMake files.

See merge request exasgd/solvers/hykkt!25

shakedregev committed:
Clean up of CMake files.

pelesh committed:
Merge branch 'maksud-refactoring' into 'develop'

Modified templates of the cuda_memory_utils functions

See merge request exasgd/solvers/hykkt!27

pelesh committed:
- Put prototypes of cuSOLVER dependent functions in a separate header file.
- Clean-up include directives.

pelesh committed:
Reduce dependencies on CUDA compiler.

pelesh committed:
Merge branch 'headers-cleanup-dev' into 'develop'

Clean up include directives in examples and update build rules

See merge request exasgd/solvers/hykkt!30

pelesh committed:
Build HyKKT as a standalone library.

pelesh committed:
Merge branch 'shared-build-fix' into 'develop'

Build HyKKT as a standalone library

Closes #23 and #24

See merge request exasgd/solvers/hykkt!32

ryandanehy committed:
Add CI/CD pipeline for HyKKT

pelesh committed:
Merge branch 'CI/CD_pipeline_develop' into 'develop'

CI/CD pipeline develop

Closes #22

See merge request exasgd/solvers/hykkt!33

ryandanehy committed:
Add pipeline to Newell for CI/CD

pelesh committed:
Merge branch 'CI_add_newell_test' into 'develop'

Ci add newell test

Closes #27

See merge request exasgd/solvers/hykkt!36

ryandanehy committed:
A better solution to the weird pipeline issue that was discovered earlier this week

pelesh committed:
Merge branch 'fix_pipeline_issue' into 'develop'

A better solution to the weird pipeline issue that was discovered earlier this week

See merge request exasgd/solvers/hykkt!38

shakedregev committed:
Expanded documentation for HyKKT. Includes documentation of individual classes in the library, as well as examples used for testing the code.

pelesh committed:
Merge branch 'README' into 'develop'

Readme

Closes #16, #17, and #19

See merge request exasgd/solvers/hykkt!26

pelesh committed:
Merge branch 'timeout-fix' into 'develop'

Increase slurm time limit from 10 mins to 60 mins.

Closes #36

See merge request exasgd/solvers/hykkt!39

ryandanehy committed:
Increase slurm time limit from 10 mins to 60 mins.

pelesh committed:
Fixing multiple warnings in HyKKT code.

pelesh committed:
Merge branch 'warnings-fix' into 'develop'

Fixing multiple warnings in HyKKT code.

See merge request exasgd/solvers/hykkt!40

shakedregev committed:
Improve allocation checks and expand comments in the code.

pelesh committed:
Merge branch 'expand_comments' into 'develop'

Expand comments and fix some allocate checks

See merge request exasgd/solvers/hykkt!42

pelesh committed:
Merge branch 'readme-update-dev' into 'develop'

Update README.md

See merge request exasgd/solvers/hykkt!43

pelesh committed:
Update README.md [skip ci]

shakedregev committed:
[skip ci] Fixed param in and out comments

pelesh committed:
[skip ci] Merge branch 'param_in_out' into 'develop'

Fixed param in and out comments

Closes #37

See merge request exasgd/solvers/hykkt!44

pelesh committed:
Define HyKKT API

pelesh committed:
Merge branch 'hykkt-api-dev' into 'develop'

Define HyKKT API

Closes #21, #25, and #26

See merge request exasgd/solvers/hykkt!34

pelesh committed:
Merge branch 'make_install' into 'develop'

install rules added for examples, lib, and include

Closes #38

See merge request exasgd/solvers/hykkt!45

ryandanehy committed:
Install rules added for examples, lib, and include

pelesh committed:
Merge branch 'clang-format' into 'develop'

added clangformat-11 testing in cmake and gitlab ci

See merge request exasgd/solvers/hykkt!46

ryandanehy committed:
added clangformat-11 testing in cmake and gitlab ci

pelesh committed:
Merge branch 'release_01' into 'develop'

Updated version, and license/disclaimers

Closes #44

See merge request exasgd/solvers/hykkt!47

ryandanehy committed:
Updated version, and license/disclaimers

ryandanehy committed:
Refactor CI and Mirror to PNNL (#1)

* Refactor CI and Mirror to PNNL

* Update push_mirror.yaml

* No Message found pipeline fix

---------

Co-authored-by: ryan.danehy@pnnl.gov <dane678@we45149.pnl.gov>

pelesh committed:
[skip ci] Minor fixes to README files (#3)

* Correct LaTeX embedding.

* Correct links in README file

cameronrutherford commented on CMakeLists.txt:
```diff
@@ -1,29 +1,40 @@
+message(STATUS "Installation Destination is ${CMAKE_INSTALL_PREFIX}")
+
+# Including clang-format cmake files to do automatic checking of formating
+include(./cmake/clang-format)
```
I think this relative path should probably be changed to an absolute path using the right CMake variables, or we should configure a module path so that this is a little cleaner for cuda and clang-format utilities (maybe we can take care of this in the linting PR) - maybe there should be an issue for the linting PR on GitHub?

pelesh replied:
Will be addressed in #6

cameronrutherford commented on CMakeLists.txt:
```diff
@@ -34,22 +45,41 @@ set(CMAKE_CUDA_SEPARABLE_COMPILATION OFF)
-                      CUDA::cudart)
+include(./cmake/FindHykktCudaLibraries)
 
 # Enable testing
 enable_testing()
 
-# TODO - use target based includes
-include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src/include
-                    ${CMAKE_CURRENT_SOURCE_DIR}/src
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src
```
This TODO should still be there...

```suggestion
# TODO - use target based includes
include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src
```

pelesh replied:
To be addressed in #6

cameronrutherford commented on buildsystem/newell-env.sh:
```diff
@@ -0,0 +1,11 @@
+
```
```suggestion
#!/bin/bash
```

cameronrutherford commented on buildsystem/deception-env.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+source /etc/profile.d/modules.sh
```
```suggestion
#!/bin/bash

source /etc/profile.d/modules.sh
```

Don't know why this was nuked - you still need bash to run module commands and scripts...

cameronrutherford commented on buildsystem/newell-env.sh:
```diff
@@ -0,0 +1,11 @@
+source /etc/profile.d/modules.sh
+module purge
+
+module use -a /qfs/projects/exasgd/src/cameron/spack/share/spack/modules/linux-centos8-power9le
+
+module load gcc/8.5.0
+module load cmake-3.23.2-gcc-8.5.0-tpplkft
```
```suggestion
module load gcc/8.5.0
module load cmake/3.19.6
```

Let's just use a system module instead of a flaky spack module

cameronrutherford commented on deception_test.sbatch:
```diff
@@ -9,7 +9,33 @@
 #SBATCH -o test.%J.out
 #SBATCH -e test.%J.err
 
+set -x
+
+# Load modules
+source /etc/profile.d/modules.sh
+
+module purge 
+module load cmake/3.19.6
+module load gcc/9.1.0
+module load cuda/11.4
+
+ulimit
+
+
+mkdir -p build
+
+rm -rf build/*
+
+cmake -B build -S . && 
+
+cmake --build build
+
 
 cd build
 ctest -VV
 
+EXIT_CODE=$?
+echo BUILD_STATUS:$EXIT_CODE
+exit $EXIT_CODE
+
```
```suggestion
MY_CLUSTER=deception ./buildsystem/build.sh
```

The scripts were equivalent...

cameronrutherford commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,117 @@
+SVC Account Cleanup:
+  stage: .pre
+  extends:
+   - .pnnl_nonhpc_tags
```
```suggestion
   - .pnnl_tags_template
```

This is the same change as in #5

cameronrutherford reviewed (changes requested):
Just did a quick review of CMake and scripts - LGTM and perhaps these changes can be prolonged for now.

Also see #5 - I made a review suggestion that applies the same 1 line change here

Commit:
Use HPC job for cleanup stage and fix PNNL CI (#5)

* Use HPC job for cleanup stage

Otherwise job doesn't have access to shared file system, and can't actually clean up jobs.

* Test failing script

* Remove debugging failure

pelesh commented:
@CameronRutherford, thank you for helpful comments. The issues you raised escaped reviews in prior PRs and are tracked in #6.
